Add hwaccess library framework

gcpin opened the pull request:
Verified:
Verified that upon setting certain access method, respective methods are getting called.
Installed the targeting-app & targeting libs on bmc image using recipe, and getCfam operation was successful.
This demonstrates the .so libraries (targeting, hwaccess and transport) are correctly installed and working as expected.

root@p11bmc:/tmp/targeting# ./targeting-app
PROC: Physical:/Sys0/Node0/MI0

//Set HwAccessMethod to Direct Access
HwBaseAccess::getCfam executed
HwBaseAccess::putCfam executed

//Set HwAccessMethod to SBEFIFO
HwBaseAccess::getCfam executed
HwBaseAccess::putCfam executed

HwSbeAccess::getScom executed
HwSbeAccess::putScom executed

//Set HwAccessMethod to Direct Access
HwDirectAccess::getScom executed
HwDirectAccess::putScom executed

targeting-app: getCfamRegister
HwBaseAccess::getCfam executed
getFSIBasePath Found FSI base path: /sys/class/fsi-master/
getcfam for addr=0x2810 value=0x80ff6007

gcpin committed:
targetsvc: Add setHwAccessMethod to set access method for a target

setHwAccessMethod shall be used by the user application to set if
the target uses DirectAccess or SBEFifo Access for subsequent calls
to getscom, putscom etc.

Also meson changes are done to use hwaccess library by targetsvc.

Signed-off-by: Gopichand Paturi <gopichandpaturi@gmail.com>

devenrao committed:
target: fix typo related to include folder

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: I82bbf5d8c590f8bdadf0aa173cfb1b84181a9e6e

devenrao committed:
target: rename target service include name

Keep variables used in meson for targetsvc library and
targetsvc include consitent

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: I3163f4670a1d3f9c0544be77463b4150e34d6e43

devenrao committed:
meson: use declared dependencies for libraries

Refactored build definitions to use `declare_dependency()` for
fdtlib, targetlib and targetsvclib.

This bundles library linkage, include paths, and transitive dependencies
into a single dep object, making it easier to consume in other targets.

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: Ic25d1a76676000a48818b36bd2f33882248f8950

devenrao committed:
targeting: use fully qualified include paths for public headers

Updated all includes to use the installed header layout under
`<targeting/...>` and `<targetsvc/...>` instead of relying on
flat local paths. This makes the code consistent with the
exported include tree from meson and avoids accidental header
resolution from the source directory.

Examples:
- `<association.H>` → `<targeting/association.H>`
- `<attributetraits.H>` → `<targeting/xmltohb/attributetraits.H>`
- `<predicateattr.H>` → `<targeting/predicates/predicateattr.H>`
- `<target_service.H>` → `<targetsvc/target_service.H>`

This ensures consumers always include headers via their public
namespace and keeps build dependencies explicit.

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: I1bf8bbe658c7a32a7d2ebc21f1213a065e1c8607

devenrao committed:
target: add low-level CFAM support

Add methods to perform low-level hardware access:

Direct access:
 - getcfam

These APIs provide the foundation for higher-level transport and
targeting operations.

Tested
root@p11bmc:/tmp# ./transport-app
Test1: GetCFAM on proc0 target
getFSIBasePath Found FSI base path: /sys/class/fsi-master/
getcfam for addr=0x2810 value=0x21704

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: Ib17986d6bb6b4b006c76ac61115d9121a31c940d

devenrao committed:
target: add low-level direct access SCOM support

Add methods to perform low-level hardware access:

Direct access:
 - getscom

These APIs provide the foundation for higher-level transport and
targeting operations.

Tested
root@p11bmc:/tmp# ./transport-app
getScom for addr=0x50001 value=0x200c084
successfully read scom value 0x200c084

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: I8f1d69756283d8435d987583d4b9d4efbe9d85ca

devenrao committed:
target: add low-level sbefifo scom support

Sbefifo access:
 - sendAndRecv

These APIs provide the foundation for higher-level transport and
targeting operations.

sbefifo getscom command
cmd [size=10]: 00 00 00 04 00 00 a2 01 00 00 00 00 00 05 00 01
value [size=20]: 84 c0 00 02 00 00 00 00 c0 de a2 01 00 00 00 00 00 00 00 03
cmd [size=16]: 00 00 00 04 00 00 a2 01 00 00 00 00 00 02 00 11
value [size=20]: 00 00 00 00 00 00 00 00 c0 de a2 01 00 00 00 00 00 00 00 03

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: I4aa48febc507d320c8d35c47416b384d952c6ec1

devenrao committed:
target: add low-level putCfam and putScom support

Add methods to perform low-level hardware access:

Direct access:
 - putCfam
 - putscom

Tested:
Test1: GetCFAM on proc0 target
getFSIBasePath Found FSI base path: /sys/class/fsi-master/
getcfam for addr=0x2810 value=0x21700
successfully read addr:0x2810 cfam value 0x21700
getFSIBasePath Found FSI base path: /sys/class/fsi-master/
putcfam for addr=0x2810 value=0x170200
successfully write addr:0x2810 cfam value 0x21700
getFSIBasePath Found FSI base path: /sys/class/fsi-master/
getcfam for addr=0x2810 value=0x21700
successfully read addr:0x2810 cfam value 0x21700
getScom for addr=0x50001 value=0x200c084
successfully read addr:0x50001 scom value 0x200c084
putScom for addr=0x50001 value=0x84c0000200000000
successfully write addr:0x50001 scom value 0x200c084
getScom for addr=0x50001 value=0x200c084
successfully read addr:0x50001 scom value 0x200c084
sbefifo getscom command
cmd [size=10]: 00 00 00 04 00 00 a2 01 00 00 00 00 00 05 00 01
value [size=20]: 84 c0 00 02 00 00 00 00 c0 de a2 01 00 00 00 00 00 00 00 03
cmd [size=16]: 00 00 00 04 00 00 a2 01 00 00 00 00 00 02 00 11
value [size=20]: 00 00 00 00 00 00 00 00 c0 de a2 01 00 00 00 00 00 00 00 03
root@p11bmc:/tmp# ./transport-app

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: Ic857790eee4424533fa449420667dcfedb7d1852

devenrao committed:
target: fix wrap-based subproject download in bitbake

Bitbake disables automatic wrap-based downloads, which caused the
build to fail with:

  ERROR: Automatic wrap-based subproject downloading is disabled

Updated the meson build to remove reliance on wrap downloads and
use proper system/Yocto-provided dependencies instead.

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: Icf7b9e4824b556f8f09b22e4ac460e6517fbcd90

devenrao committed:
targeting: fix shared library packaging Yocto

Yocto do_package_qa was failing with:

ERROR: -dev package pdbg-dev contains non-symlink .so
    '/usr/lib/libtargeting.so'

The issue was that unversioned .so files were being installed as real
ELF objects instead of symlinks. Updated meson.build to specify
`version` and `soversion` in `library()` calls, ensuring that:

  - Actual ELF shared objects are installed with full version
    (e.g. libfoo.so.1.0.0)
  - Runtime symlinks (libfoo.so.1) are created
  - Development symlinks (libfoo.so) are created for -dev packages

This resolves Yocto QA errors and aligns library packaging with
standard conventions.

Signed-off-by: Marri Devender Rao <devenrao@in.ibm.com>
Change-Id: If2607ef9c5af3d02603de17417d6e0b77c8e9dd7

gcpin committed:
hwaccess: Create hwAccess classes and methods

This commit adds a new hwAccess library which will be the interface
to access hwTransport layer APIs that do direct_fsi/sbe_fifo access
via kernel read/write.

DirectAccess & SbeAccess shall route calls to respective transport
interfaces/layers.

Signed-off-by: Gopichand Paturi <gopichandpaturi@gmail.com>

gcpin committed:
hwaccess: Create hwAccess Interface for user applications

Added 5 interface methods
getcfam, putcfam, getscom, putscom, gethwaccessptr

External applications
1. Need to utilize hwAccessIntf static functions for cfam/scom access
2. User Application/library should set HwAccessMethod explicitly for a target,
   without which access shall fail with an exception.
3. It is the responsibility of the user to ensure the target on which these
   interface methods are called have HW_ACCESS_METHOD and HW_ACCESS_PTR
   attributes, otherwise exception will be thrown for incompatible targets.

Signed-off-by: Gopichand Paturi <gopichandpaturi@gmail.com>

gcpin committed:
hwaccess: Add meson changes to build hwaccess library

hwaccess will be a shared object (.so) library.

Signed-off-by: Gopichand Paturi <gopichandpaturi@gmail.com>

gcpin committed:
targeting: Fix setVolatileAttr template function

The function signature is different from the intended function
signature, modifying to the intended version to make it compilable.

Signed-off-by: Gopichand Paturi <gopichandpaturi@gmail.com>

gcpin committed:
hwaccess: Integrate hwTransport layer for Direct Access methods

This commit integrates the transport layer functionality to call into
the transport layer for getCfam, putCfam, getScom & putScom functions.

Signed-off-by: Gopichand Paturi <gopichandpaturi@gmail.com>

gcpin committed:
libfdt: Move include headers to fdt directory and add versioning

Prevents include header conflict with other libfdt header files if they
are installed via pdbg etc. Also add versioning to the library to
facilitate installation through recipe.

Signed-off-by: Gopichand Paturi <gopichandpaturi@gmail.com>

deepakala-k commented on target/hwaccess/hwAccessIntf.C:
```diff
@@ -0,0 +1,84 @@
+
+namespace hwaccess
+{
+int HwAccessIntf::getCfamRegisters(TargetPtr_t target, uint32_t addr, uint32_t& val)
```
This method gets a single register value, so better to name it getCfamRegister. (remove the s) and similarly for all the other methods

deepakala-k commented on target/hwaccess/hwBaseAccess.C:
```diff
@@ -0,0 +1,26 @@
+{
+    if(target == nullptr)
+    {
+        return -1;
```
It would be better to have debug traces on why we are returning an error. Similar comment for all the places where we return with an error.

devenrao replied:
why the ccommit message does not have commit id? did you not install hooks applicable every where

deepakala-k commented on target/hwaccess/hwAccessIntf.C:
```diff
@@ -0,0 +1,84 @@
+        return hwAccessPtr->getCfam(target, addr, val);
+    }
+
+    return -1;
```
In the HwBaseAccess and the other access classes, we return as soon the target pointer is nullptr, here if the hwAccessPtr is nullptr, it comes to the end and returns. Can we have a similar approach everywhere

deepakala-k commented on target/targetsvc/target_service.C:
```diff
@@ -67,4 +68,14 @@ TargetPtr TargetService::getTopLevelTarget() const
+    uintptr_t accessPtr =
+            reinterpret_cast<uintptr_t>(hwaccess::HwAccessIntf::getHwAccessPtr(accessMethod));
+
+    target->setAttr<ATTR_HW_ACCESS_PTR>(static_cast<HwAccessPtrType>(accessPtr));
```
Where do we handle the exception that is thrown if the setAttr and getAttr failed? Is the caller expected to handle it? Should that be added to the function header?

devenrao replied:
caller needs to handle the exceptions

gcpin replied:
will add to function header, it will be upon caller to handle the exceptions

devenrao commented on target/hwaccess/hwBaseAccess.H:
```diff
@@ -0,0 +1,24 @@
+    virtual ~HwBaseAccess() = default;
+
+public:
+    int getCfam(TargetPtr_t target, uint32_t addr, uint32_t& val);
```
better to pass ConstTargetPtr_t for all the methods as we do not change any internals of target in these methods

deepakala-k replied:
Yeah better to change it to const reference to avoid creating temp variables

devenrao commented on target/hwaccess/hwBaseAccess.H:
```diff
@@ -0,0 +1,24 @@
+
+#include <target.H>
+#include <attributeenums.H>
+#include <attributetraits.H>
```
use complete path like targeting/target.H
targeting/xmltohb/attributeenums.H
targeting/xmltohb/attributetraits.H

applicable every where

gcpin replied:
targeting library is including all the include directories like xmltohb etc in the meson file. I think this is better way to include, as the build system will search the file name in the include directories of the library. If I give absolute path , and if we modify xmltohb directory name or move file to some other directory, we will have to modify this path in all libraries. If we keep it this way build system will search in the include directories even if the name changes.

gcpin replied:
Made changes to include root directory i.e target as include header directory for all libs. Now absolute paths from this position can be used anywhere to access any header files across libs and within. For headers within same directory, double quotes shall be used.

devenrao commented on target/hwaccess/hwBaseAccess.H:
```diff
@@ -0,0 +1,24 @@
+#include <target.H>
+#include <attributeenums.H>
+#include <attributetraits.H>
+#include <hwAccessIntf.H>
```
where is this header file not part of this commit?

gcpin replied:
this change was left out when i was splitting the commit. will update.

devenrao commented on target/hwaccess/hwAccessIntf.C:
```diff
@@ -0,0 +1,84 @@
+#include <hwAccessIntf.H>
```
this also use "hwAccessIntf.H>

also these file names are not as per openbmc guidelines it got to be
hw_access_intf.H and similarily others do check guidelines on file names

devenrao commented on target/hwaccess/hwAccessIntf.C:
```diff
@@ -0,0 +1,84 @@
+    if (hwAccessPtr != nullptr)
+    {
+        return hwAccessPtr->getCfam(target, addr, val);
+    }
```
in else case we can add assert to indicate it should never happen

gcpin replied:
modified to check for nullptr and log error. i think we cannot assert (never fails) for hwAccessPtr as the typecast happens at run time and getAttr will throw exception if the accessPtr is not available.

devenrao commented on target/hwaccess/hwAccessIntf.H:
```diff
@@ -0,0 +1,21 @@
+
+#include <target.H>
+#include <attributeenums.H>
+#include <attributetraits.H>
```
use complete install path for include files

devenrao commented on target/hwaccess/hwAccessIntf.H:
```diff
@@ -0,0 +1,21 @@
+    static int putCfamRegisters(TargetPtr_t target, uint32_t addr, uint32_t val);
+    static int getScomRegisters(TargetPtr_t target, uint64_t addr, uint64_t& val);
+    static int putScomRegisters(TargetPtr_t target, uint64_t addr, uint64_t val);
+    static void* getHwAccessPtr(HwAccessMethod& accessMethod);
```
once every thing is done add documentation doxygen for these methods

devenrao commented on target/hwaccess/meson.build:
```diff
@@ -0,0 +1,22 @@
+)
+
+# Export public headers into /usr/include/hwaccess
+install_headers('hwAccessIntf.H', subdir: 'hwaccess')
```
define dependency so that consumers can use, looks like u had done it in next commit,

gcpin replied:
yes, done in next commits.

devenrao commented on target/hwaccess/meson.build:
```diff
@@ -0,0 +1,22 @@
+  '.',
+)
+
+hwaccess_lib = shared_library(
```
should this be static library?

gcpin replied:
As per our design discussions we decided to keep all the libraries as .so (shared) for now.

devenrao commented on target/libfdt/meson.build:
```diff
@@ -49,7 +49,8 @@ install_headers(
     'libfdt.h',
     'libfdt_env.h',
-  )
+  ),
+  subdir: 'fdt'
```
we need to change the existing code where we do #include<libfdt.H> to #include<fdt/libfdt.H> then

gcpin replied:
Changes may  not be needed, all the libraries that were dependent on libfdt compiled without an error after i made this change also.

gcpin committed:
hwaccess: Rename files to comply with openbmc guidelines

Signed-off-by: Gopichand Paturi <gopichandpaturi@gmail.com>

gcpin committed:
target: Make all libraries consistent with include headers

1. Files in current directory need to be included using double quotes
2. All other files to be included must use absolute path from target/
3. Included a specific directory in include path as exception if the
   directory holds auto-generated code. Example: targeting/xmltohb

This brings consistency and makes portability easier when decoupling.

Signed-off-by: Gopichand Paturi <gopichandpaturi@gmail.com>

gcpin committed:
hwaccess: Add doxygen documentation

Signed-off-by: Gopichand Paturi <gopichandpaturi@gmail.com>

devenrao commented on target/libfdt/libfdt.h:
```diff
@@ -6,8 +6,8 @@
-#include <fdt.h>
+#include "libfdt_env.h"
+#include "fdt.h"
 
```
this is not our code so undo these changes

devenrao commented on target/targeting/xmltohb/targAttrIdToName.C:
```diff
@@ -1,4 +1,4 @@
-#include <targAttrIdToName.H>
+#include "targAttrIdToName.H"
 
```
this is generated code so do not make changes here

gcpin replied:
if we are to include only parent directory (as we discussed to keep things cleaner), we would not be able to use angular brackets as include path to xmltohb wont be available and double quotes need to be used. In interest of time I am adding xmltohb in include path to remove this change. We can discuss on this further later.

devenrao reviewed (commented):
will approve after code is rebased so far looks good addressed earlier review comments
